sds/src/scoped_ruleset/mod.rs

```diff
@@ -199,21 +215,53 @@ where
         // Sanitize the segment and push it
         self.sanitized_segments_until_node.push(segment.sanitize());
 
+        let true_positive_rules_count = if self.should_keywords_match_event_paths {
+            let mut current_sanitized_path = self
+                .sanitized_segments_until_node
+                .iter()
+                .filter_map(|sanitized_segment| {
+                    sanitized_segment.as_ref().map(|seg| Some(seg.clone()))
+                })
+                .fold(String::new(), |mut a, b| {
+                    let b_str = b.expect(
+                        "In the filter_map above, we make sure to filter out the None variants",
+                    );
+                    a.reserve(b_str.len() + 1);
```

Comment:
`String::reserve` is quite expensive. This is what actually does the allocation, so it would be good to minimize this as much as possible. Luckily, we already have a list of all of the segments, so you can just do a quick sum over the size of the segments to get the total length, and create a string of the correct capacity with only a single allocation with `String::with_capacity`.

I went ahead and implemented this to make sure it's actually better, and I'm seeing a decent perf improvement

```rust
let mut total_len: usize = self
                .sanitized_segments_until_node
                .iter()
                .flatten()
                .map(|x| x.len() + 1)
                .sum();
            if total_len > 0 {
                // subtract the len of the last link char that isn't needed
                total_len -= 1;
            }

            let mut current_sanitized_path = String::with_capacity(total_len);
            for (i, segment) in self
                .sanitized_segments_until_node
                .iter()
                .cloned()
                .flatten()
                .enumerate()
            {
                if i != 0 {
                    current_sanitized_path.push(UNIFIED_LINK_CHAR);
                }
                current_sanitized_path.push_str(segment.as_ref());
            }
```

<img width="636" alt="image" src="https://github.com/user-attachments/assets/45a6f265-7c36-4792-aa0b-0792262257ab">

Reply:
[Improve current_sanitized_path creation algorithm](https://github.com/DataDog/dd-sensitive-data-scanner/pull/97/commits/6f75d45d277ec11cb4631c5a9e8eb7aec29c94e1)